runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -597,16 +597,16 @@ public synchronized Instant getEarliestTimerTimestamp() {
       Instant earliest = THE_END_OF_TIME.get();
       for (NavigableSet<TimerData> timers : processingTimers.values()) {
         if (!timers.isEmpty()) {
-          earliest = INSTANT_ORDERING.min(timers.first().getTimestamp(), earliest);
```

Comment:
I wonder if this is sufficient. The NavigableSet is ordered by timer firing time, not by outputTimestamp. This means that the logic to just look at the first timer (which was correct before we had holds) is probably no longer enough. we probably need to iterate over the set (or we need to have a second data structure to hold the watermark holds and take the minimum of that data structure).

Reply:
@reuvenlax  added a new function to calculate the earliest outputTimestamp from NavigableSet.